pkg/network/http/http_stats.go

```diff
@@ -26,53 +45,87 @@ type RequestStats [5]struct {
 	// Note: every time we add a latency value to the DDSketch below, it's possible for the sketch to discard that value
 	// (ie if it is outside the range that is tracked by the sketch). For that reason, in order to keep an accurate count
 	// the number of http transactions processed, we have our own count field (rather than relying on DDSketch.GetCount())
-	count     int
-	latencies *ddsketch.DDSketch
+	Count     int
+	Latencies *ddsketch.DDSketch
+
+	// This field holds the value (in milliseconds) of the first HTTP request
+	// in this bucket. We do this as optimization to avoid creating sketches with
+	// a single value. This is quite common in the context of HTTP requests without
+	// keep-alives where a short-lived TCP connection is used for a single request.
+	FirstLatencySample float64
 }
 
 // CombineWith merges the data in 2 RequestStats objects
+// newStats is kept as it is, while the method receiver gets mutated
 func (r *RequestStats) CombineWith(newStats RequestStats) {
-	for i := 0; i < 5; i++ {
-		r[i].count += newStats[i].count
+	for i := 0; i < len(r); i++ {
+		statusClass := 100 * (i + 1)
+		switch newStats[i].Count {
+		case 0:
+			// No data to be merged
+			continue
+		case 1:
+			// The other bucket has a single latency sample, so we "manually" add it
+			r.AddRequest(statusClass, newStats[i].FirstLatencySample)
+			continue
+		default:
+			// The other bucket (newStats) has multiple samples and therefore a DDSketch object
+			// We first ensure that the bucket we're merging to has a DDSketch object
+			if r[i].Latencies == nil {
+				// TODO: Consider calling Copy() on the other sketch instead
+				if err := r.initSketch(i); err != nil {
+					continue
+				}
+
+				// If we had a latency sample in this bucket we now add it to the DDSketch
+				if r[i].Count == 1 {
+					r[i].Latencies.Add(r[i].FirstLatencySample)
+				}
+			}
 
-		if r[i].latencies == nil {
-			r[i].latencies = newStats[i].latencies
-		} else if newStats[i].latencies != nil {
-			err := r[i].latencies.MergeWith(newStats[i].latencies)
+			// Finally merge both sketches
+			r[i].Count += newStats[i].Count
+			err := r[i].Latencies.MergeWith(newStats[i].Latencies)
 			if err != nil {
-				log.Debugf("Error merging HTTP transactions: %v", err)
+				log.Debugf("error merging http transactions: %v", err)
 			}
```

Comment:
I know this isn't new code, but when would we expect a failure when merging two sketches?

Reply:
err, I don't know 😬 
I'll follow up on that

Reply:
MergeWith can return an error if the index mappings don't match: https://github.com/DataDog/sketches-go/blob/c9a8f6a89b352df76f2e49e07d977a9ce163d560/ddsketch/ddsketch.go#L184-L189

This _shouldn't_ happen, but since it's technically possible I added the error handling just do be safe 🤷